client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java

```diff
@@ -0,0 +1,320 @@
+        
+        private final String name;
+        private final InputStream in;
+        private final ReplyProcessor callback;
+        private final Thread thread;
+        private final AppendBatchSizeTracker batchSizeTracker;
+        private final AtomicBoolean stop = new AtomicBoolean(false);
+
+        public ConnectionReader(String name, InputStream in, ReplyProcessor callback,
+                                AppendBatchSizeTracker batchSizeTracker) {
+            this.name = name;
+            this.in = in;
+            this.callback = callback;
+            this.thread = THREAD_FACTORY.newThread(this);
+            this.batchSizeTracker = batchSizeTracker;
+        }
+        
+        public void start() {
+            thread.start();
+        }
+        
+        public void run() {
+            IoBuffer buffer = new IoBuffer();
+            while (!stop.get()) {
+                try {
+                    WireCommand command = readCommand(in, buffer);
+                    if (command instanceof WireCommands.DataAppended) {
+                        WireCommands.DataAppended dataAppended = (WireCommands.DataAppended) command;
+                        batchSizeTracker.recordAck(dataAppended.getEventNumber());
+                    }
+                    try {
+                        callback.process((Reply) command);
+                    } catch (Exception e) {
+                        callback.processingFailure(e);
+                    }
+                } catch (SocketException e) {
+                    if (e.getMessage().equals("Socket closed")) {
+                        log.info("Closing TcpConnection.Reader because socket is closed.");
+                    } else {
+                        log.warn("Error in reading from socket.", e);
+                    }
+                    stop();
+                } catch (EOFException e) {
+                    log.info("Closing TcpConnection.Reader because end of input readched.");
+                    stop();
+                } catch (Exception e) {
+                    log.warn("Error processing data from from server " + name, e);
+                    stop();
+                }
+            }
+        }
+
+        @VisibleForTesting
+        static WireCommand readCommand(InputStream in, IoBuffer buffer) throws IOException {
+            ByteBuf header = buffer.getBuffOfSize(in, 8);
+
+            int t = header.getInt(0);
+            WireCommandType type = WireCommands.getType(t);
+            if (type == null) {
+                throw new InvalidMessageException("Unknown wire command: " + t);
+            }
+
+            int length = header.getInt(4);
+            if (length < 0 || length > WireCommands.MAX_WIRECOMMAND_SIZE) {
+                throw new InvalidMessageException("Event of invalid length: " + length);
+            }
+
+            ByteBuf payload = buffer.getBuffOfSize(in, length);
```

Comment:
the same here.
we can ask Netty to take the buffer from its pool (as we already know the length we also won't fall into buffer reallocations).
Netty will be able to manage direct memory for us.
and EnhancedByteBufInputStream will call release on "close()" so Netty will be able to recycle the buffer/memory.

It would be nice to make EnhancedByteBufInputStream recyclable in order to save another object allocation at every command unmashalling.

Reply:
I don't want to add further dependency on netty functionality here given that I am trying to remove it all together: https://github.com/pravega/pravega/issues/4923

It may make sense to add a dependency on the same buffer manager we use on the server side which we benchmarked as being faster than netty's version. That change however is large enough in it's own right that I don't want to lump it into this PR which is already quite large. I agree the above is not perfect it shows large perf improvements over the current master.

Reply:
Agreed